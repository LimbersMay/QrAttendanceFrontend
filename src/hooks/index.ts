export * from './useForm';
export * from './useUiStore';
export * from './useAuthStore';
export * from './useCheckAuth';
export * from './useGroupStore';
export * from './useQrCodeStore';
export * from './useRegistryStore';
export * from './useQrAttendanceStore';