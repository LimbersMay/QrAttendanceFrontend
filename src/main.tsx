import React from 'react'
import ReactDOM from 'react-dom/client'
import {QrAttendanceApp} from "./QrAttendanceApp";

ReactDOM.createRoot(document.getElementById('root') as HTMLElement).render(
  <React.StrictMode>
    <QrAttendanceApp />
  </React.StrictMode>,
)
