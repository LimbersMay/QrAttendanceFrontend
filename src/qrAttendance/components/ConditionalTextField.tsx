import {TextField} from "@mui/material";
import {ChangeEventHandler} from "react";

export const ConditionalTextField = ({ onChange, condition, name, value, styles }: { onChange: ChangeEventHandler, condition: boolean, name: string, value: any, styles?: Object}) => {
    return (
        condition
            ? <TextField variant={'standard'} sx={styles} onChange={onChange} name={name} value={value}></TextField>
            : value
    )
}