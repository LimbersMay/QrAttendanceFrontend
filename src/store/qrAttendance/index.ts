export * from "./group";
export * from "./qrCode";
export * from "./registry";
