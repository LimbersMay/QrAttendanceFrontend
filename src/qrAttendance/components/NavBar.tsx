import React from "react";
import {AppBar, Avatar, Box, Grid, IconButton, Menu, MenuItem, Toolbar, Tooltip, Typography} from "@mui/material";
import {MenuOutlined} from "@mui/icons-material";
import {useAuthSlice} from "../../hooks/useAuthSlice";

export const NavBar = ({ handleDrawerToggle, drawerWidth = 240 }: {handleDrawerToggle: any, drawerWidth: number }) => {

    const { startLogout, displayName } = useAuthSlice();

    const [anchorElUser, setAnchorElUser] = React.useState<null | HTMLElement>(null);

    const handleOpenUserMenu = (event: React.MouseEvent<HTMLElement>) => {
        setAnchorElUser(event.currentTarget);
    };

    const handleCloseUserMenu = () => {
        setAnchorElUser(null);
    };

    const handleLogout = async () => {
        await startLogout();
    }

    return (
        <AppBar
            position='fixed'
            sx={{
                width: {sm: `calc(100% - ${drawerWidth}px)`},
                ml: {sm: `${drawerWidth}px`}
            }}
        >
            <Toolbar>
                <IconButton
                    onClick={handleDrawerToggle}
                    color='inherit'
                    edge='start'
                    sx={{
                        mr: 2,
                        display: {sm: 'none'}
                    }}
                >
                    <MenuOutlined />
                </IconButton>

                <Grid
                    container
                    direction='row'
                    justifyContent='space-between'
                    alignItems='center'
                >
                    <Typography variant='h6' noWrap component='div'>QrAttendance</Typography>

                    <Box sx={{ flexGrow: 0 }}>
                        <Tooltip title="Open settings">
                            <IconButton onClick={handleOpenUserMenu} sx={{ p: 0 }}>
                                <Avatar alt={`${displayName}`} />
                            </IconButton>
                        </Tooltip>
                        <Menu
                            sx={{ mt: '45px' }}
                            id="menu-appbar"
                            anchorEl={anchorElUser}
                            anchorOrigin={{
                                vertical: 'top',
                                horizontal: 'right',
                            }}
                            keepMounted
                            transformOrigin={{
                                vertical: 'top',
                                horizontal: 'right',
                            }}
                            open={Boolean(anchorElUser)}
                            onClose={handleCloseUserMenu}
                        >
                            <MenuItem onClick={handleCloseUserMenu}>
                                <Typography textAlign="center">Profile</Typography>
                            </MenuItem>
                            <MenuItem onClick={handleCloseUserMenu}>
                                <Typography textAlign="center">My account</Typography>
                            </MenuItem>
                            <MenuItem onClick={handleLogout}>
                                <Typography textAlign="center">Logout</Typography>
                            </MenuItem>
                        </Menu>
                    </Box>
                </Grid>
            </Toolbar>
        </AppBar>
    )
}
