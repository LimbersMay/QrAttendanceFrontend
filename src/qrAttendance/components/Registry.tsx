import {useState} from "react";

import {
    TextField,
    TableCell,
    IconButton,
    TableRow
} from "@mui/material";
import {Delete, Edit, Save} from "@mui/icons-material";

import {useForm} from "../../hooks/useForm";

interface Registry {
    id: string,
    date: string,
    name: string,
    sourname: string,
    lastname: string
}

export const Registry = ({registryRow }: { registryRow: Registry}) => {

    const [isRowEditing, setIsRowEditing] = useState(false);
    const {formState, onInputChange} = useForm(registryRow);

    const { name, sourname, lastname, date } = formState;

    const handleEdit = () => {
        setIsRowEditing(true);
    }

    const handleSave = () => {
        setIsRowEditing(false);
    }

    const handleDelete = () => {
        console.log('delete');
    }

    return (
        <TableRow>
            <TableCell component="th" scope="row">
                {
                    isRowEditing
                        ? <TextField sx={{width: '140px'}} onChange={onInputChange} name="date" value={date}></TextField>
                        : date
                }
            </TableCell>
            <TableCell>{name}</TableCell>
            <TableCell align="center">{sourname}</TableCell>
            <TableCell align="center">
                { lastname}
            </TableCell>
            <TableCell align="center">
                {
                    isRowEditing
                        ? <IconButton onClick={handleSave}><Save/></IconButton>
                        : <IconButton onClick={handleEdit}><Edit /></IconButton>
                }
                <IconButton onClick={handleDelete}>
                    <Delete />
                </IconButton>
            </TableCell>
        </TableRow>
    )
}