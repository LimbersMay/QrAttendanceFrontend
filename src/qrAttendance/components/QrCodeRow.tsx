import {useState} from "react";

import {IconButton, MenuItem, Select, SelectChangeEvent, TableCell, TableRow} from "@mui/material";
import KeyboardArrowUpIcon from "@mui/icons-material/KeyboardArrowUp";
import KeyboardArrowDownIcon from "@mui/icons-material/KeyboardArrowDown";
import {Save} from "@mui/icons-material";

import {useForm} from "../../hooks/useForm";
import {ConditionalTextField} from "./ConditionalTextField";
import {QrCode} from "../interfaces";
import {useAppDispatch} from "../../store";
import { startDeleteQrCodeWithDependencies, startUpdateQrCode} from "../../store/qrAttendance";
import {QrCodeMenuOptions} from "./QrCodeMenuOptions";

export const QrCodeRow = ({
           qrCodeRow,
           handleOpenSubTable,
           open,
    registriesLength
}:{ qrCodeRow: QrCode, handleOpenSubTable: Function, open: boolean, registriesLength: number }) => {

    const dispatch = useAppDispatch();

    const [isEditing, setIsEditing] = useState<boolean>(false);
    const [isEnable, setIsEnable] = useState<boolean>(qrCodeRow.enabled);

    const {formState, onInputChange} = useForm(qrCodeRow);
    const {name, date} = formState;

    const onSelectChange = (event: SelectChangeEvent) => {
        if (event.target.value === 'YES') return setIsEnable(true);
        setIsEnable(false);
    }

    const handleEdit = () => {
        setIsEditing(true);
    }

    const handleSave = () => {
        setIsEditing(false);
        dispatch(startUpdateQrCode({
            ...qrCodeRow,
            ...formState,
            enabled: isEnable
        }));
    }

    const handleDelete = () => {
        dispatch(startDeleteQrCodeWithDependencies(qrCodeRow.id));
    }

    const handleShow = () => {

    }

    const handleDownload = () => {

    }

    return (
        <TableRow sx={{'& > *': {borderBottom: 'unset'}}}>
            <TableCell>
                <IconButton
                    aria-label="expand row"
                    size="small"
                    onClick={() => handleOpenSubTable(!open)}
                >
                    {open ? <KeyboardArrowUpIcon/> : <KeyboardArrowDownIcon/>}
                </IconButton>
            </TableCell>
            <TableCell component="th" scope="row">
                <ConditionalTextField
                    name="name"
                    value={name}
                    onChange={onInputChange}
                    condition={isEditing}
                    styles={{width: '120px'}}
                />
            </TableCell>
            <TableCell align="center">
                {registriesLength}
            </TableCell>
            <TableCell align="center">
                <ConditionalTextField
                    name="date"
                    value={date}
                    onChange={onInputChange}
                    condition={isEditing}
                    styles={{width: '100px'}}
                />
            </TableCell>
            <TableCell align="center">
                {
                    isEditing
                        ? <Select variant={'standard'} value={isEnable ? 'YES' : 'NO'} onChange={onSelectChange}>
                            <MenuItem value="YES">Yes</MenuItem>
                            <MenuItem value="NO">No</MenuItem>
                        </Select>
                        : isEnable ? 'YES' : 'NO'
                }
            </TableCell>
            <TableCell align="center">
                {
                    isEditing
                        ? <IconButton onClick={handleSave}><Save/></IconButton>
                        : <QrCodeMenuOptions
                            handleEdit={handleEdit}
                            handleDelete={handleDelete}
                            handleShow={handleShow}
                            handleDownload={handleDownload}
                        />
                }

            </TableCell>
        </TableRow>
    )
}