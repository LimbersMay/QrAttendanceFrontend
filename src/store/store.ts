import {AnyAction, configureStore, ThunkAction} from '@reduxjs/toolkit';
import {authSlice} from "./auth";
import {groupSlice, qrCodeSlice, registrySlice} from "./qrAttendance";
import socketMiddleware from "./middlewares/socketMiddleware";
export const store = configureStore({
    reducer: {
        auth: authSlice.reducer,
        group: groupSlice.reducer,
        qrCode: qrCodeSlice.reducer,
        registry: registrySlice.reducer
    },
    middleware: [socketMiddleware]
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;

export type AppThunk<ReturnType = void> = ThunkAction<
    ReturnType,
    RootState,
    unknown,
    AnyAction
>
