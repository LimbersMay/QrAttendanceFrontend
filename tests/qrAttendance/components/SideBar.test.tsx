import {configureStore} from "@reduxjs/toolkit";
import {authSlice} from "../../../src/store/auth";
import {groupSlice} from "../../../src/store/qrAttendance";
import {authenticatedState} from "../../fixtures/authStates";
import {
    withActiveGroupAndGroupsState
} from "../../fixtures/qrAttendanceStates";
import {render, screen} from "@testing-library/react";
import {Provider} from "react-redux";
import {ThemeProvider} from "@mui/material";
import {purpleTheme} from "../../../src/theme";
import {MemoryRouter} from "react-router-dom";
import {SideBar} from "../../../src/qrAttendance/components";

const store = configureStore({
    reducer: {
        auth: authSlice.reducer,
        group: groupSlice.reducer,

    },
    preloadedState: {
        auth: {...authenticatedState},
        group: {...withActiveGroupAndGroupsState},
    }
})

describe('Tests for <SideBar />', () => {
    beforeEach(() => jest.clearAllMocks());

    test('should render the component successfully', async () => {
        const {container} = await render(
            <Provider store={store}>
                <ThemeProvider theme={purpleTheme}>
                    <MemoryRouter>
                        <SideBar drawerWidth={280} mobileOpen={false} handleDrawerToggle={() => {}}/>
                    </MemoryRouter>
                </ThemeProvider>
            </Provider>
        );

        expect(container).toMatchSnapshot();
    });

    test('should show the user name', () => {

        render(
            <Provider store={store}>
                <ThemeProvider theme={purpleTheme}>
                    <MemoryRouter>
                        <SideBar drawerWidth={280} mobileOpen={false} handleDrawerToggle={() => {}}/>
                    </MemoryRouter>
                </ThemeProvider>
            </Provider>
        );

        expect(screen.getByText(`${authenticatedState.displayName}`)).toBeTruthy();
    });

    test('should render the groups in the store', () => {
        render(
            <Provider store={store}>
                <ThemeProvider theme={purpleTheme}>
                    <MemoryRouter>
                        <SideBar drawerWidth={280} mobileOpen={false} handleDrawerToggle={() => {}}/>
                    </MemoryRouter>
                </ThemeProvider>
            </Provider>
        );

        const groups = screen.getAllByTestId('group-item');
        expect(groups.length).toBe(withActiveGroupAndGroupsState.groups.length);
    });
});
