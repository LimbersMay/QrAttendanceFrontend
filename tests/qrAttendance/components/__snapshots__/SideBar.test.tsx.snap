// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <SideBar /> should render the component successfully 1`] = `
<div>
  <nav
    class="MuiBox-root css-lksrek"
  >
    <div
      class="MuiDrawer-root MuiDrawer-docked css-1fk4axq-MuiDrawer-docked"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiDrawer-paper MuiDrawer-paperAnchorLeft MuiDrawer-paperAnchorDockedLeft css-12i7wg6-MuiPaper-root-MuiDrawer-paper"
      >
        <div
          class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
        >
          <div
            class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-8u39c-MuiTypography-root"
          >
            Limbers
          </div>
        </div>
        <hr
          class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
        />
        <ul
          class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-i0cfj8-MuiGrid-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-fullWidth css-ttzjjx-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              New Group
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <li
            class="MuiListItem-root MuiListItem-gutters css-vymmwx-MuiListItem-root"
          >
            <div
              class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-1y3heoy-MuiButtonBase-root-MuiListItemButton-root"
              role="button"
              tabindex="0"
            >
              <div
                class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="TurnedInNotIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M17 3H7c-1.1 0-1.99.9-1.99 2L5 21l7-3 7 3V5c0-1.1-.9-2-2-2zm0 15-5-2.18L7 18V5h10v13z"
                  />
                </svg>
              </div>
              <div
                class="MuiGrid-root MuiGrid-container css-1vam7s3-MuiGrid-root"
              >
                <div
                  class="MuiListItemText-root css-10tin8d-MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
                  >
                    System engineering
                  </span>
                </div>
                <div
                  class="MuiListItemText-root css-tlelie-MuiListItemText-root"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
                  >
                    Invalid Date
                  </p>
                </div>
              </div>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </div>
          </li>
          <li
            class="MuiListItem-root MuiListItem-gutters css-vymmwx-MuiListItem-root"
          >
            <div
              class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-1y3heoy-MuiButtonBase-root-MuiListItemButton-root"
              role="button"
              tabindex="0"
            >
              <div
                class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="TurnedInNotIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M17 3H7c-1.1 0-1.99.9-1.99 2L5 21l7-3 7 3V5c0-1.1-.9-2-2-2zm0 15-5-2.18L7 18V5h10v13z"
                  />
                </svg>
              </div>
              <div
                class="MuiGrid-root MuiGrid-container css-1vam7s3-MuiGrid-root"
              >
                <div
                  class="MuiListItemText-root css-10tin8d-MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
                  >
                    English A group
                  </span>
                </div>
                <div
                  class="MuiListItemText-root css-tlelie-MuiListItemText-root"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
                  >
                    Invalid Date
                  </p>
                </div>
              </div>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </div>
          </li>
        </ul>
      </div>
    </div>
  </nav>
</div>
`;
